test/marvelus/function_test.zls

```diff
@@ -0,0 +1,87 @@
+    # (=>(and (= v z) (= z (+ x y)) (> y 0.0) (> x 0.0)) (and(>= v x) (>= v y))
+   Constraints:
+   - `x` and `y` must be positive floats.
+   - The return value `v` should satisfy the constraints `x <= v` and `y <= v`.
+*)
+let add (x : {v:float|v>0.0} ) (y : {v:float|v>0.0} ) : { v:float | x <= v && y <= v } =
+  let z = x +. y in
+  z
+
+(* Example 2: Sub Function (pass*)
+(*
+   The sub function takes two float values, `x` and `y`,
+   where `x` is a positive float and `y` is a non-negative float.
+   It returns a float value `t` such that `t <= x`.
+   This function should pass the refinement verification.
+    # (=>(and (= t z) (= z (- x y)) (> x 0.0) (>= y 0.0)) (and(<= t x))
+   Constraints:
+   - `x` must be a positive float.
+   - `y` must be a non-negative float.
+   - The return value `t` should satisfy the constraint `t <= x`.
+*)
+let sub (x : {w:float|w>0.0} ) (y : {r:float|r>0.0} ) : { t:float | t <= x } =
+  let z = x -. y in
+  z
+
+(* Example 3: Multiply Function (fail) *)
+(*
+   The multiply function takes two float values, `x` and `y`,
+   and returns a float value `z` such that `z >= x` and `z >= y`.
+   This function is expected to fail the refinement verification.
+    # (=>(and (= r z) (= z (/*/ x y)) (> y 0.0) (> x 0.0)) (or(>= r x) (<=r x) (>= r y) (<=r x))
+   Constraints:
+   - `x` and `y` are integers.
+   - The return value `z` should satisfy the constraints `z >= x` and `z >= y`.
+*)
+let multiply (x : {w:float|w>0} ) (y : {v:float|v>0}) : {r:float | r >= x && r >= y } =
+  let z = x *. y in
+  z
+
+(* Example 4: Division Function (fail) *)
+(*
+   The divide function takes two non-zero integer values, `x` and `y`,
+   and returns a float value `z` such that `z = x / y`.
+   This function should pass the refinement verification.
+    # (=>(and (= r z) (= z (/ x y))  (>= y 0.0) (= x 0.0)) (or(<= r x) (>= r x))
+   Constraints:
+   - `x` and `y` are non-zero integers.
+   - The return value `z` is the division of `x` by `y`.
+*)
+let divide (x : {v:float | v > 0 }) (y : {v:float | v > 0 }) : {r: float | r <= x } =
+  let z = x /. y in
+  z
+
+(* Example 5: Division Function with negative values (fail) *)
+(*
+   The divide function takes two negative integer values, `x` and `y`,
+   and returns a float value `z` such that `z = x / y`.
+   This function should fail the refinement verification.
+    # (=>(and (= r z) (= z (/ x y))  (not (= y 0.0)) (not (= x 0.0))) (or (<= r x) (>= r x))
+   Constraints:
+   - `x` and `y` can be negative integers.
+   - The return value `z` is the division of `x` by `y`.
+*)
+let divide_pos_neg (x : {v:float | v > 0 || v < 0 }) (y : {v:float | v > 0  || v < 0}) : {r: float | r <= x} =
+  let z = x /. y in
+  z
+
+(* Entry Point *)
+(*
+   The node function serves as the entry point for the test.
+   It calls the `add` , `sub` and `multiply` functions with sample arguments,
+   prints the results, and returns unit ().
+*)
+let node main () =
```

Comment:
In a few of the outputs, I can see a "42" magic number (see below). Where are those coming from? 
This usually complicates more complicated proofs and would be good to fix whenever we see them.

```
vc_gen_expression 42
Operator call: print_float
vc_gen_expression 2.0
Operator call: add
map2 error is here!!
name of function:add
length of ref_fun.argument_list:2
# of Arguments in arg_list:2 arguments:2
Proving constraint: (not (=> (and (= 2.0 x) (= 3.0 y)) (and (> y 0.0) (> x 0.0))))
Passed
```

Reply:
Looks like most of the 42's are coming from `prove_function`, line 1073, `z3refinement.ml`(not modified by this PR). Perhaps something hasn't been fully implemented there? It seems like all functions get turned into `42` regardless of whether or not they exist in the environment. There's also a TODO there: "check if arguments obey constraints". Does this PR address this TODO comment?

Reply:
I'm also noticing `Eseq` causing 42's to appear, which combined with the `42` issue with the functions above, results in a bunch of magic numbers appearing. May not be relevant to include in the current PR but still needs to be addressed.

Reply:
I commented out the function application on the tests and only looked at function declarations for this review. The only issue I see (which is more like a design choice than an issue), is that the output type of a function is tied to a variable in the verification condition (v = z & x <= v & y <= v), instead of applying the substitution of the variable directly in the boolean constraint (x <= z & y <= z). But I think this works assuming the return value is a single variable (not pairs, with each element typed separately).

`
let add (x : {v:float|v>0.0} ) (y : {v:float|v>0.0} ) : { v:float | x <= v && y <= v } =
  let z = x +. y in
  z
`
`
Proving constraint: (let ((a!1 (=> (and (= v z) (= z (+ x y)) (> y 0.0) (> x 0.0))
               (and (<= x v) (<= y v)))))
  (not a!1))
`